### PR TITLE
State: Type notice actions and their options

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

As part of our work in #48408, we've been reduxifying a bunch of notices in the checkout area. Much of it being typed already, it came to our attention (thanks @sirbrillig for bringing it up) that those can cause errors because they're not typed. 

This PR introduces basic types for our notice action creators and their options, in order to make using them in TS code easier, and to help with decrypting the possible options that can be passed.

#### Testing instructions
* Verify build still works well.
* Smoke test notices throughout Calypso.
* Verify all tests still pass.

### CODE REVIEW

**Comment by sirbrillig** [comment]
Great work here! This will be really helpful when using the notice action creators in a TS context to make sure the options are correct (and prevent the errors that appear when using them otherwise).

File: client/state/notices/types.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+export interface NoticeOptions extends BaseNoticeOptions {
+	id?: NoticeId;
+}
+
+// Notice options as they're returned by the action creator
+export interface NoticeActionOptions extends BaseNoticeOptions {
+	noticeId: NoticeId;
+	status: NoticeStatus;
+	text: NoticeText;
+}
```
Comment (sarayourfriend) [inline client/state/notices/types.ts:46]: Could this interface be turned into a generic one where the status is specific? Could that render any benefit or would it just complicate things without benefiting it?

File: client/state/notices/actions.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+export const infoNotice: NoticeActionCreator = ( text, noticeOptions ) =>
+	createNotice( 'is-info', text, noticeOptions );
+export const warningNotice: NoticeActionCreator = ( text, noticeOptions ) =>
+	createNotice( 'is-warning', text, noticeOptions );
+export const plainNotice: NoticeActionCreator = ( text, noticeOptions ) =>
+	createNotice( 'is-plain', text, noticeOptions );
+
+// Higher-order action creator: modify the wrapped creator to return actions with the
+// `notices.skip` meta, so that it's ignored by the notices middleware.
+export const withoutNotice = ( actionCreator ) => ( ...args ) =>
```
Comment (sarayourfriend) [inline client/state/notices/actions.ts:56]: Is the type missing for this one?

File: client/state/notices/types.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	text: NoticeText,
+	noticeOptions?: NoticeOptions
+) => NoticeAction;
+
+// Notice removal action creator function
+export type NoticeRemovalActionCreator = (
+	noticeId: NoticeId
+) => {
+	noticeId: NoticeId;
+	type: string;
```
Comment (sarayourfriend) [inline client/state/notices/types.ts]: Could `type` be more specific here?

File: client/state/notices/types.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	noticeId: NoticeId;
+	status: NoticeStatus;
+	text: NoticeText;
+}
+
+// Notice action as returned by the action creator
+export type NoticeAction = {
+	type: string;
+	notice: NoticeActionOptions;
+};
```
Comment (sarayourfriend) [inline client/state/notices/types.ts]: Can this also be generic on the `type` property?

File: client/state/notices/types.ts (hunk 1 above)
Comment (tyxla) [reply client/state/notices/types.ts:46]: I don't see any benefit coming from making these generic by status - while we benefit from the fact that we type all those options, the rest of the properties are the same, and I believe that by providing a more general type for all notices (vs separating them by status) we allow for better flexibility. Also, TBH, I don't see much value in such generic typing from a consumer point of view. We can always update that if we eventually end up needing granularity, though.

File: client/state/notices/actions.ts (hunk 2 above)
Comment (tyxla) [reply client/state/notices/actions.ts:56]: Yup 👍  Added in https://github.com/Automattic/wp-calypso/pull/49461/commits/ccb4b4949071fcbe1401d698c2726690b0a1723d

Note that this type likely needs to live somewhere else (in `/state/types.ts` for example) but I'm leaving that for another time. Typing all of `state` and figuring out types that fit well for all of Calypso's state feels like a separate task.

Comment (tyxla) [reply client/state/notices/actions.ts:56]: Scratch that, I'll need to improve this bit.

File: client/state/notices/types.ts (hunk 3 above)
Comment (tyxla) [reply client/state/notices/types.ts]: Redux `type` is just a string anyway. We currently don't type our action types in any way. What would you suggest instead?

File: client/state/notices/types.ts (hunk 4 above)
Comment (tyxla) [reply client/state/notices/types.ts]: See https://github.com/Automattic/wp-calypso/pull/49461#discussion_r567926811 for my rationale on that.

File: client/state/notices/actions.ts (hunk 2 above)
Comment (tyxla) [reply client/state/notices/actions.ts:56]: @saramarcondes I'll have to type `extendAction` too to get that done properly. Given this one (`withoutNotice`) is used very little throughout the codebase, I think I'll leave it for another time if you don't mind.

File: client/state/notices/types.ts (hunk 3 above)
Comment (sarayourfriend) [reply client/state/notices/types.ts]: I didn't realize we didn't type our actions at all. Nothing to do here in that case.

For what it's worth, typically the type of actions in Redux can by strictly typed as a union of types where they all share the same `type` property with strict values. An example can be found here: https://github.com/ItsJonQ/g2/blob/d8d4c53ad3eb6bbac50a61fe854dd47c088f4912/packages/components/src/Accordion/useAccordion.js#L77

File: client/state/notices/actions.ts (hunk 2 above)
Comment (sarayourfriend) [reply client/state/notices/actions.ts:56]: Sounds good to me.

File: client/state/notices/types.ts (hunk 3 above)
Comment (tyxla) [reply client/state/notices/types.ts]: Actually, you're bringing a good point here. We don't need to allow all action types, but we only care about a couple of them. Fixed in c60d5d1089693b1724d1d722cee4c48d551a39e2

**Review by sarayourfriend — APPROVED** [review]
Looks good from my perspective.

**Comment by tyxla** [comment]
I'm going to ship this as-is, but if you run into further trouble and any of those needs refining, let me know @sirbrillig, I'm happy to refine them further. 

Thanks both! 🙌

**Pull request merged**